src/Nerdbank.Streams/MultiplexingStream.Channel.cs

```diff
@@ -916,17 +994,28 @@ private async Task AutoCloseOnPipesClosureAsync()
 
             private void Fault(Exception exception)
             {
-                if (this.TraceSource?.Switch.ShouldTrace(TraceEventType.Critical) ?? false)
+                this.mxStreamIOReader?.CancelPendingRead();
+
+                // If the channel has already been disposed then only cancel the reader
+                lock (this.SyncObject)
                 {
-                    this.TraceSource!.TraceEvent(TraceEventType.Critical, (int)TraceEventId.FatalError, "Channel Closing self due to exception: {0}", exception);
+                    if (this.isDisposed)
+                    {
+                        return;
+                    }
                 }
 
+                // Set the faulting exception to the current exception if it isn't already set
                 lock (this.SyncObject)
                 {
```

Comment:
Try to avoid entering the same lock you just exited. Why not join these two blocks?

Reply:
I initially had some other piece of code between the two and thus just left that logic in place.